File: App/Pages/Topics/topic.js
S.topic = {
    edited: false,
    dropzone: null,

    load: function () {
        $('#btnaddsection').on('click', function () { S.topic.buttons.addSection('section', 0); });
        $('.btn-select-all-images').on('click', S.topic.media.toggleSelect);
        $('.btn-gallery-toggle').on('click', S.topic.media.toggleGallery);
        S.topic.media.refresh();

        //init dropzone library
        Dropzone.autoDiscover = false;

        S.topic.dropzone = new Dropzone(document.body, {
            url: '/Topics/Upload/?topicId=' + S.topic.id,
            previewsContainer: ".topic-media .upload-list",
            clickable: ".topic-media .btn-upload",
            paramName: 'file',
            maxFilesize: 4,
            uploadMultiple: true,
            thumbnailWidth:100,
            thumbnailHeight: 80,
            parallelUploads: 1
                    , init: function () {
                        this.on('sending', function () {
                            $('.topic-media .dropzone').addClass('uploading');
                        });

                        this.on('complete', function (file) {
                            this.removeFile(file);
                        });

                        this.on('queuecomplete', function () {
                            S.ajax.post('/api/Topics/SaveUpload', {}, S.ajax.callback.inject);
                            $('.topic-media .dropzone').addClass('uploaded').removeClass('uploading');
                        });
                    }
        });
    },

    buttons: {
        editTopic: function (className) {
            console.log(className);
            $('.accordion.' + className + ' .preview, .' + className + ' .btn-cancel').hide();
            $('.accordion.' + className + ' .edit').show();
            autosize.update($('#' + className + '-content'));
        },

        previewTopic: function (className) {
            var title = $('.accordion.' + className + ' .txt-title').val();
            var content = $('.accordion.' + className + ' .txt-content').val();
            var htm = marked(content
                    .replace(/\n{3,}/g, '\n\n!b!\n\n')
                    .replace(/\n{3,}/g, '\n\n!b!\n\n')
                ).replace(/\<p\>\!b\!\<\/p\>/g, '<br/>');
            $('.accordion.' + className + ' .edit, .preview .nopreview').hide();
            $('.accordion.' + className + ' .preview, .preview .ispreview').show();
            $('.accordion.' + className + ' .section-contents')[0].innerHTML = htm;
            $('.accordion.' + className + ' > .title').html(title);
        },

        addSection: function (group, index) {
            var sections = $('.topic-section');
            var section = $('.' + group + index + ' .topic-section');
            var count = sections.length;
            var element = '.' + group + (index + 1);
            var after = false;
            if (section[0] == sections[sections.length - 1]) { after = true; element = '.' + group + index; }
            var options = { element: element, after: after, title: "New Section", content: "", index: index, count: count };
            S.ajax.post('/api/Topics/NewTopicSection', options, function () {
                S.ajax.callback.inject(arguments[0]);
            });
        },

        removeSection: function (group, index) {
            if (confirm('Do you really want to delete this section?')) {
                $('.section' + index).remove();
                S.ajax.post('/api/Topics/RemoveSection', {index: index}, S.ajax.callback.inject);
            }
            
        },

        saveChanges: function (groupName) {
            if (S.topic.edited == false) { return; }
            var i = 1;
            var e = 0;
            var data = new Array();
            var obj;
            var sections = $('.topic-section');
            var section;
            var id = '';
            var title = '';
            $('#btnsavechanges').hide();
            S.topic.edited = false;
            for (var s = 0; s < sections.length; s++) {
                section = sections[s];
                id = section.className.replace('topic-section id-' + groupName, '');
                title = $(section).find('.txt-title').val();
                data.push({
                    title: title,
                    content: $(section).find('.txt-content').val(),
                    id: id
                });
                //update title
                $('.' + groupName + id + ' > .title').html(title);
            }
            S.ajax.post('/api/Topics/SaveTopic', { sections: JSON.stringify(data) }, function () { });
        }
    },

    texteditor: {
        keyDown: function (e) {
            if (S.topic.edited == false) {
                $('#btnsavechanges').show();
                S.topic.edited = true;
                autosize.update($(e));
            }
        },

        autoSize: function () {
            var ta = document.querySelector('textarea');
            
            //remove any existing autosizes
            var evt = document.createEvent('Event');
            evt.initEvent('autosize:destroy', true, false);
            ta.dispatchEvent(evt);

            //add new autosizes to all textareas on the page
            autosize($('textarea'));
        }
    },

    media: {
        refresh: function () {
            //updates media items
            $('.media-list .chk').off().on('click', S.topic.media.check);
        },

        check: function () {
            //select a media item
            var chk = $(this).find('input[type=checkbox]');
            if (chk.is(":checked") == true) {
                $(this).addClass('checked');
            } else {
                $(this).removeClass('checked');
            }
            S.topic.media.selected();
        },

        selected: function () {
            //check if any other images are selected
            if ($('.media-list .chk input[type=checkbox]:checked').length > 0) {
                $('.btn-delete-selected-images').show();
            } else {
                $('.btn-delete-selected-images').hide();
            }
        },

        toggleSelect: function(){
            if ($('.media-list .chk input[type=checkbox]:checked').length > 0) {
                //deselect
                $('.media-list .chk input[type=checkbox]:checked').prop("checked", false);
                $('.media-list .chk').removeClass('checked');
            } else {
                //select all
                $('.media-list .chk input[type=checkbox]').prop("checked", true);
                $('.media-list .chk').removeClass('checked').addClass('checked');
            }
            S.topic.media.selected();
        },

        toggleGallery: function () {
            var list = $('.media-list');
            if(list.hasClass('gallery')){
                list.removeClass('gallery');
            } else {
                list.addClass('gallery');
            }
            
        }
    }
};

S.topic.load();

File: App/Pages/Topics/Topic.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Collector.Pages
{
    public class Topic : Page
    {
        public Topic(Core CollectorCore) : base(CollectorCore)
        {
        }

        public override string Render(string[] path, string body = "", object metadata = null)
        {
            if (!CheckSecurity()) { return AccessDenied(); }

            //load dashboard interface
            var dashboard = new Partials.Dashboard(S);

            //add custom menu to dashboard
            dashboard.AddMenuItem("btnaddsection", "+ New Section", "");
            dashboard.AddMenuItem("btnsavechanges", "Save Changes", "");

            //load variables for page content
            var accordion = new Partials.Accordion(S);
            var query = new Query.Topics(S.Server.sqlConnection);
            var service = new Services.Topics(S);
            var html = new StringBuilder();
            var topic = query.GetDetails(int.Parse(path[1]));
            var topicPath = "/content/topics/" + topic.hierarchy.Replace(">", "/") + "/";
            var topicFile = topicPath + topic.topicId + ".json";

            //render topic summary list item into an accordion
            html.Append(accordion.Render("Topic", "topic-summary", service.LoadTopicListItem(topic).Render()) + "\n");

            //render topic media
            html.Append(accordion.Render("Images", "topic-media", GetTopicMediaList(topic)) + "\n");

            //load topic sections from cache or JSON file
            var sections = GetTopicSectionsFromJSON(topicFile);

            //render each topic section
            var i = 0;
            foreach (var section in sections)
            {
                i++;
                html.Append(GetTopicEditableAccordion(topic.title, section.title, section.content, "section", "section" + i, topic.topicId, i, true, i == 1 ? true : false, true));
            }

            //add page resources
            dashboard.scripts += scripts;
            dashboard.AddScript("/js/utility/dropzone.js");
            dashboard.AddScript("/js/utility/marked.js");
            dashboard.AddScript("/js/utility/autosize.min.js");
            dashboard.AddScript("/js/pages/topics/topic.js");
            dashboard.AddCSS("/css/utility/dropzone.css");
            dashboard.AddCSS("/css/pages/topics/topic.css");

            //finally, render page
            return dashboard.Render(path, html.ToString(), metadata);
        }

        private List<Query.Models.TopicSection> GetTopicSectionsFromJSON(string topicFile)
        {
            if (File.Exists(S.Server.MapPath(topicFile)))
            {
                return (List<Query.Models.TopicSection>)S.Util.Serializer.ReadObject(S.Server.LoadFileFromCache(topicFile), typeof(List<Query.Models.TopicSection>));
            }
            return new List<Query.Models.TopicSection>();
        }

        private string GetTopicEditableAccordion(string topicTitle, string title, string content, string groupName, string className, int topicId, int index, bool whiteBg = false, bool expanded = true, bool editMode = false, bool isNew = false)
        {
            var accordion = new Partials.Accordion(S);

            //register javascript events for the editable form
            scripts += "<script language=\"javascript\">" +  
                    (!editMode ? (content != "" ? "S.topic.buttons.previewTopic('" + className + "');" : "") : "") +
                    "$('." + className + "-btn-edit').on('click', function(){S.topic.buttons.editTopic('" + className + "');});" +
                    "$('." + className + "-btn-preview').on('click ', function(){S.topic.buttons.previewTopic('" + className + "');});" +
                    "$('." + className + "-btn-newsection').on('click', function(){S.topic.buttons.addSection('"+ groupName + "'," + index + ");});" +
                    "$('." + className + "-btn-remove').on('click', function(){S.topic.buttons.removeSection('" + groupName + "'," + index + ");});" +
                    "$('.topic-section input, .topic-section textarea').off().on('keydown', function(e){S.topic.texteditor.keyDown(e.target);});" +
                    "$('.btn-savechanges').off().on('click', function(){S.topic.buttons.saveChanges('" + groupName + "');});" +
                    "S.topic.texteditor.autoSize();" + 
                "</script>";

            //render an accordion with two tabs, one for editing a topic sectionl, the other for previewing
            return accordion.Render(title, className,
                    "<div class=\"topic-section id-section" + index + "\">" + 
                        //live preview
                        "<div class=\"preview\"" + (editMode ? " style=\"display:none\"" : "") + ">" +
                            "<div class=\"nopreview\"" + (content == "" ? "" : " style=\"display:none;\"") + ">" + 
                                "<div class=\"title\">Start writing in new section for the topic '" + topicTitle + "'.</div>" +
                                "<a href=\"javascript:\" class=\"button green " + className + "-btn-edit\">Edit Section</a>" + 
                            "</div>" +
                            "<div class=\"ispreview\"" + (content != "" ? "" : " style=\"display:none;\"") + ">" +
                                "<div class=\"section-contents\">" + "</div>" + //content + "</div>" +
                                "<div class=\"buttons\">" +
                                    "<a href=\"javascript:\" class=\"button green left " + className + "-btn-edit\">Edit</a>" +
                                    "<a href=\"javascript:\" class=\"button blue left " + className + "-btn-newsection\">+ New Section</a>" +
                                    "<a href=\"javascript:\" class=\"button green right btn-savechanges\" style=\"display:none;\">Save Changes</a>" +
                                "<a href=\"javascript:\" class=\"button left " + className + "-btn-remove\">Remove</a>" +
                                "</div>" +
                            "</div>" +
                        "</div>" +
                        //editable form used for markdown
                        "<div class=\"edit\"" + (editMode ? "" : " style=\"display:none\"") + ">" +
                            "<div class=\"row column label\">Section Title</div>" +
                            "<div class=\"row column\"><input type=\"text\" class=\"txt-title\" value=\"" + title.Replace("\"","\\\"") + "\"></div>" +
                            "<div class=\"row column label\">Content <span class=\"right\"><a href=\"https://guides.github.com/features/mastering-markdown/\" target=\"_blank\">Markdown</a></span></div>" +
                            "<div class=\"row column\"><textarea class=\"txt-content\">" + content + "</textarea></div>" +
                            "<div class=\"row column buttons\">" +
                                "<a href=\"javascript:\" class=\"button green left " + className + "-btn-preview\">Preview</a>" +
                                "<a href=\"javascript:\" class=\"button blue left " + className + "-btn-newsection\">+ New Section</a>" +
                                "<a href=\"javascript:\" class=\"button green left btn-savechanges\" style=\"display:none;\">Save Changes</a>" +
                                "<a href=\"javascript:\" class=\"button left " + className + "-btn-remove\">Remove</a>" +
                            "</div>" +
                        "</div>" +
                    "</div>"
                , expanded, whiteBg);
        }

        private string GetTopicMediaList(Query.Models.Topic topic, bool listOnly = false)
        {
            //display list of images
            var media = topic.media.Split(',', StringSplitOptions.RemoveEmptyEntries);
            var htm = new StringBuilder();
            if (listOnly == false)
            {
                htm.Append("<div class=\"media-list\">");
            }
            for (var x = 0; x < media.Length; x++)
            {
                if (media[x] != "")
                {
                    htm.Append(
                        "<div class=\"img img-" + x + "\">" +
                            "<div class=\"chk\"><input type=\"checkbox\"/></div>" +
                            "<img src=\"/topics/" + topic.hierarchy.Replace(">", "/") + "/sm_" + media[x] + "\"/></div>");
                }
            }
            if (listOnly == false)
            {
                htm.Append("</div>" +
                    "<div class=\"img-details\">" +

                    "</div>");

                //add drop zone for uploading images
                htm.Append(
                    "<div class=\"upload-list\"></div>" +
                        "<div class=\"buttons\">" +
                            "<a href=\"javascript:\" class=\"button green btn-upload left\">Upload Images</a>" +
                            "<a href=\"javascript:\" class=\"button btn-select-all-images left\">Select All / None</a>" +
                            "<a href=\"javascript:\" class=\"button btn-delete-selected-images red left\">Delete Selected</a>" +
                            "<a href=\"javascript:\" class=\"button blue btn-gallery-toggle right\">View As Gallery</a>" +
                        "</div>" +
                    "<div class=\"dropzone\">" +
                    "<span class=\"drop-here\">Drop Images Here</span>" +
                    "</div>");
            }
            return htm.ToString();
        }
    }
}
